src/renderer/util.ts

```diff
@@ -183,29 +184,42 @@ type ValType<T> =
   | React.ChangeEvent<HTMLInputElement>
   | (Record<string, unknown> & { value?: T; checked?: T });
 
+type NestedType<T, P> = P extends `${infer K}.${infer NestedKey}`
+  ? K extends keyof T
+    ? NestedType<T[K], NestedKey>
+    : undefined
+  : P extends keyof T
+  ? T[P]
+  : undefined;
+
 export function useSetting<
   T extends Record<string, Jsonifiable>,
   D extends keyof T,
   K extends Extract<keyof T, string>,
-  F extends T[K] | undefined,
+  F extends NestedType<T, P> | T[K] | undefined,
+  P extends `${K}.${string}` | K,
 >(
   settings: SettingsManager<T, D>,
-  key: K,
+  key: P,
   fallback?: F,
 ): {
-  value: K extends D
-    ? NonNullable<T[K]>
-    : F extends null | undefined
-    ? T[K] | undefined
-    : NonNullable<T[K]> | F;
-  onChange: (newValue: ValType<T[K]>) => void;
+  value: NonNullable<NestedType<T, P>>;
+  onChange: (newValue: ValType<NonNullable<NestedType<T, P>>>) => void;
```

Comment:
You can't just wrap the whole thing in `NonNullable`, it needs to match how it was done before.